### PR TITLE
help IDE support

### PR DESCRIPTION
**Pull request by mwhipple**
This shifts custom checks to be done through subclassing rather than the current jamming on the metaclass, and passing the library to using with the `@DelegatesTo` annotation provides checking and completion support in IDEA.

A lingering hack is jamming an `Idator` into the Checkers class to make sure that the assorted `define` methods are also picked up by IDEA. I don't yet know a better way to do that: it may require a custom type checking extension but there are some other things to address before the entire DSL could be type checked

### CODE REVIEW

**Comment by mwhipple** [comment]
This also includes initial work for encapsulating the traversal step which could help with a bunch of readability/ease of use enhancements

**Comment by mwhipple** [comment]
HipChat test

File: src/main/groovy/val/Checkers.groovy [hunk 1]
```diff
@@ -16,6 +16,10 @@ import java.util.regex.Pattern
  */
 class Checkers {
 
+    //Hack to provide completion
+    @Delegate()
+    Idator dummyMethods = new Idator()
```
Comment (mwhipple) [inline src/main/groovy/val/Checkers.groovy]: I'm planning on defining the Idator methods in an interface and then have this object be an implementation that is an angry sentinel and just throws...something, probably `UnsupportedOperationException`

File: src/main/groovy/val/Idator.groovy [hunk 2]
```diff
@@ -18,75 +18,126 @@ package val
...
-                    overrides.resultKey   ?: resultKey]
-        this.class.newInstance(*args)
+    // Runtime metaprogramming to dispatch to Checkers
+
+    // Will presently most likely have issues if Maps are non-mold parameters
+    // This could be avoided through some form of counting or an annotation
+    // The solution is deferred until the problem arises
+
+    /**
+     * Create an invocation of the provided function which with
```
Comment (samperman) [inline src/main/groovy/val/Idator.groovy]: with which

**Comment by mwhipple** [comment]
yeah...documentation is the most important missing piece.  I think I'm going to start working on porting most of the code to Java to be JVM portable (without dependencies) and will document as I go (and the fully configured test is a good idea, the tests in general could be beefed up with a lot of example uses).

The primary change in this PR is the passing of the Checkers to `using` which allows the use of the `@Delegates` annotations.  That enables code analysis which provides the IDE support. So the change isn't really how you'd use it, but how much help you'll get from tools while using it.

**Pull request merged**